packages/design-tokens/src/utils.ts

```diff
@@ -0,0 +1,160 @@
+import colorString from "color-string"
+import camelToKebab from "lodash.kebabcase"
+import { DeepMapObjectLeafs, Theme } from "./types"
+
+/**
+ * This allows you to map the leaf nodes of an object, and you're provided the path to that leaf as well as the value as parameters to your mapper function.
+ * This function was build to support mapping theme values to their respective CSS variable identifiers.
+ * For example:
+ * ```ts
+ * mapLeafsOfObject({
+ *  one: {
+ *    two: 4
+ *  }
+ * }, (path, value) => `${path.join(".")}: ${value}`)
+ * ```
+ * Results in:
+ * ```
+ * {
+ *    one: {
+ *      two: "one.two: 4"
+ *    }
+ * }
+ * ```
+ */
+export const mapLeafsOfObject = <
+  Obj extends Record<string | number, unknown>,
+  Value
+>(
+  object: Obj,
+  mapper: (pathToLeaf: string[], value: unknown) => Value
+): DeepMapObjectLeafs<Obj, Value> => {
+  const recurser = <O extends Record<string | number, unknown>>(
+    currentPath: string[],
+    obj: O
+  ): DeepMapObjectLeafs<O, Value> => {
+    const handleEntry = (key: string, value: unknown) => {
+      const pathToKey = [...currentPath, key]
+      if (typeof value === "object" && value !== null && value !== undefined) {
+        return recurser(pathToKey, value as Record<string | number, unknown>)
+      }
+      return mapper(pathToKey, value)
+    }
+    return Object.entries(obj).reduce(
+      (acc, [nextKey, nextValue]) => ({
+        ...acc,
+        [nextKey]: handleEntry(nextKey, nextValue),
+      }),
+      {} as DeepMapObjectLeafs<O, Value>
+    )
+  }
+  return recurser([], object)
+}
+
+const objectPathToCssVarIdentifier = (path: string[]) =>
+  `--${path.map(camelToKebab).join("-")}`
+
+export const objectPathToCssVarReference = (path: string[]) =>
+  `var(${objectPathToCssVarIdentifier(path)})`
+
+export const cssVariableThemeNamespace = "kz-var" as const
+/**
+ * This function will convert the leaf values of a theme to a value like `var(--parent1key-parent2key-leafkey)` - a CSS variable with an identifier that represents it's hierarchy within the object you provided.
+ * One caveat though: if a key is suffixed with `-default` it will leave the value in place. This was implemented for consumer regression purposes, i.e. so that consumers had a fallback when CSS variables weren't feasible.
+ * Example:
+ * ```
+ * {
+ *  varkz: {
+ *      color: {
+ *          wisteria: "#ff0011"
+ *      }
+ *  }
+ * }
+ * ```
+ * Transforms into:
+ * ```
+ * {
+ *  varkz: {
+ *      color: {
+ *          wisteria: "var(--kz-color-wisteria)"
+ *      }
+ *  }
+ * }
+ * ```
+ */
+
+export const makeCSSVariableTheme = (theme: Theme) =>
+  mapLeafsOfObject({ [cssVariableThemeNamespace]: theme }, (path, value) =>
+    path[path.length - 1].endsWith("-default")
+      ? value
+      : objectPathToCssVarReference(path)
+  )
+
+/**
+ * Use this to generate an object containing `${key}: value`, `${key}-default: value`, and `${key}-rgb-params: rgb(r, g, b)` if the value is a color.
```

Comment:
I wonder if the presence of "default" suffix might confuse people. This will break lots of things, but maybe the trade-off is worth it?